File: clients/client-csharp/Vexilla.Client/Scheduler.cs
using System;

namespace Vexilla.Client
{
    public class Scheduler
    {
        public static bool IsScheduleFeatureActive(Feature feature)
        {
            return IsScheduleActive(feature.Schedule, feature.ScheduleType);
        }

        public static bool IsScheduleActive(Schedule schedule, string scheduleType)
        {
            return IsScheduleActiveWithNow(schedule, scheduleType, DateTimeOffset.Now);
        }

        public static bool IsScheduleActiveWithNow(Schedule schedule, string scheduleType, DateTimeOffset now)
        {
            switch (scheduleType)
            {
                case ScheduleType.Empty:
                    return true;
                case ScheduleType.Global:
                case ScheduleType.Environment:

                    var isInDateRange = false;

                    var startDateTimeOffset = DateTimeOffset.FromUnixTimeMilliseconds(schedule.Start);

                    var startOfStartDate = new DateTimeOffset(
                        startDateTimeOffset.Year,
                        startDateTimeOffset.Month,
                        startDateTimeOffset.Day,
                        0,
                        0,
                        0,
                        startDateTimeOffset.Offset
                    );

                    var endDateTimeOffset = DateTimeOffset.FromUnixTimeMilliseconds(schedule.End);

                    var endOfEndDate = new DateTimeOffset(
                        endDateTimeOffset.Year,
                        endDateTimeOffset.Month,
                        endDateTimeOffset.Day,
                        23,
                        59,
                        59,
                        endDateTimeOffset.Offset
                    );

                    if (now < startOfStartDate || now > endOfEndDate) return false;

                    var startTime = DateTimeOffset.FromUnixTimeMilliseconds(schedule.StartTime);
                    var endTime = DateTimeOffset.FromUnixTimeMilliseconds(schedule.EndTime);


                    switch (schedule.TimeType)
                    {
                        case ScheduleTimeType.None:
                            return true;

                        case ScheduleTimeType.StartEnd:
                            var startDateTime = new DateTimeOffset(
                                startDateTimeOffset.Year,
                                startDateTimeOffset.Month,
                                startDateTimeOffset.Day,
                                startTime.Hour,
                                startTime.Minute,
                                startTime.Second,
                                startTime.Millisecond,
                                startDateTimeOffset.Offset
                            );

                            var endDateTime = new DateTimeOffset(
                                endDateTimeOffset.Year,
                                endDateTimeOffset.Month,
                                endDateTimeOffset.Day,
                                endTime.Hour,
                                endTime.Minute,
                                endTime.Second,
                                endTime.Millisecond,
                                endDateTimeOffset.Offset
                            );

                            return now > startDateTime && now < endDateTime;

                        case ScheduleTimeType.Daily:

                            var zeroDay = DateTimeOffset.FromUnixTimeMilliseconds(0);
                            var nowTimestamp = now.ToUnixTimeMilliseconds();

                            var todayZeroTimestamp = new DateTimeOffset(
                                now.Year,
                                now.Month,
                                now.Day,
                                0,
                                0,
                                0,
                                0,
                                now.Offset
                            ).ToUnixTimeMilliseconds();

                            var zeroedStartTimestamp = new DateTimeOffset(
                                zeroDay.Year,
                                zeroDay.Month,
                                zeroDay.Day,
                                startTime.Hour,
                                startTime.Minute,
                                startTime.Second,
                                startTime.Millisecond,
                                now.Offset
                            ).ToUnixTimeMilliseconds();

                            var zeroedEndDateTimeOffset = new DateTimeOffset(
                                zeroDay.Year,
                                zeroDay.Month,
                                zeroDay.Day,
                                endTime.Hour,
                                endTime.Minute,
                                endTime.Second,
                                endTime.Millisecond,
                                now.Offset
                            );
                            var zeroedEndTimestamp = zeroedEndDateTimeOffset.ToUnixTimeMilliseconds();
                            var zeroedEndTimestampPlusDay = zeroedEndDateTimeOffset.AddDays(1).ToUnixTimeMilliseconds();

                            var startTimestamp = todayZeroTimestamp + zeroedStartTimestamp;
                            var endTimestamp = todayZeroTimestamp + zeroedEndTimestamp;
                            if (zeroedStartTimestamp > zeroedEndTimestamp)
                                endTimestamp = todayZeroTimestamp + zeroedEndTimestampPlusDay;

                            return nowTimestamp > startTimestamp && nowTimestamp < endTimestamp;
                    }

                    break;

                default: return false;
            }


            return false;
        }
    }
}

File: clients/elixir/lib/Vexilla/scheduler.ex
defmodule Scheduler do
  @moduledoc false

  use Timex
  require ScheduleType
  require TimeType
  require FeatureType

  @doc """
  Checks if a feature's schedule is active

  ## Examples

    iex> Scheduler.is_scheduled_feature_active(Scheduler.get_example_feature())
    true

  """
  def is_scheduled_feature_active(feature) do
    is_schedule_active(feature["schedule"], feature["scheduleType"])
  end

  def is_schedule_active(schedule, schedule_type, datetime \\ Timex.now()) do
    schedule_type_empty = ScheduleType.empty()
    schedule_type_global = ScheduleType.global()
    schedule_type_environment = ScheduleType.environment()

    case schedule_type do
      ^schedule_type_empty ->
        true

      schedule_type when schedule_type in [schedule_type_global, schedule_type_environment] ->
        start_date = Timex.from_unix(schedule["start"], :millisecond)
        start_of_start_date = Timex.beginning_of_day(start_date)

        end_date = Timex.from_unix(schedule["end"], :millisecond)
        end_of_end_date = Timex.end_of_day(end_date)

        time_type_none = TimeType.none()
        time_type_start_end = TimeType.start_end()
        time_type_daily = TimeType.daily()

        start_time = Timex.from_unix(schedule["startTime"], :millisecond)
        end_time = Timex.from_unix(schedule["endTime"], :millisecond)

        is_date_between =
          Timex.between?(datetime, start_of_start_date, end_of_end_date, inclusive: true)

        case {schedule["timeType"], is_date_between} do
          {^time_type_none, true} ->
            true

          {^time_type_start_end, true} ->
            start_day_with_start_time =
              Timex.set(start_date,
                hour: start_time.hour,
                minute: start_time.minute,
                second: start_time.second,
                microsecond: start_time.microsecond
              )

            end_day_with_start_time =
              Timex.set(end_date,
                hour: end_time.hour,
                minute: end_time.minute,
                second: end_time.second,
                microsecond: end_time.microsecond
              )

            Timex.between?(datetime, start_day_with_start_time, end_day_with_start_time,
              inclusive: true
            )

          {^time_type_daily, true} ->
            zero_day = Timex.from_unix(0, :millisecond)
            timestamp = Timex.to_unix(datetime) * 1000

            start_of_today =
              datetime
              |> Timex.beginning_of_day()
              |> Timex.to_unix()
              |> Kernel.*(1000)

            zeroed_start_timestamp =
              Timex.set(zero_day,
                hour: start_time.hour,
                minute: start_time.minute,
                second: start_time.second,
                microsecond: start_time.microsecond
              )
              |> Timex.to_unix()
              |> Kernel.*(1000)

            zeroed_end =
              Timex.set(zero_day,
                hour: end_time.hour,
                minute: end_time.minute,
                second: end_time.second,
                microsecond: end_time.microsecond
              )

            zeroed_end_timestamp = Timex.to_unix(zeroed_end) * 1000

            start_timestamp = start_of_today + zeroed_start_timestamp
            end_timestamp = start_of_today + zeroed_end_timestamp

            if zeroed_start_timestamp > zeroed_end_timestamp do
              timestamp >= start_timestamp || timestamp <= end_timestamp
            else
              timestamp >= start_timestamp && timestamp <= end_timestamp
            end

          _ ->
            false
        end

      _ ->
        false
    end
  end

  if Mix.env() == :test do
    def get_example_feature() do
      now = Timex.now()

      %{
        "name" => "foo",
        "featureId" => "bar",
        "seed" => 0,
        "value" => 0,
        "schedule" => %{
          "start" => Timex.to_unix(Timex.subtract(now, Duration.from_days(2))) * 1000,
          "end" => Timex.to_unix(Timex.add(now, Duration.from_days(2))) * 1000,
          "timeType" => TimeType.start_end(),
          "startTime" => Timex.to_unix(Timex.subtract(now, Duration.from_hours(2))) * 1000,
          "endTime" => Timex.to_unix(Timex.add(now, Duration.from_hours(2))) * 1000
        },
        "featureType" => FeatureType.toggle(),
        "scheduleType" => ScheduleType.global()
      }
    end
  end
end
